### ISSUE TITLE
Local Cran repo doesn t work with rsconnect deployApp (even when Repository: ... is included in Description)

### ISSUE THREAD

**Issue by pommedeterresautee** [rstudio/rsconnect#238]
I want to deploy to rstudio-connect.
One of my package is on a private Github repo.
It has in its Description `Repository: myCRAN`
I have set a drat repo and declared it in options()
```R
repos <- c("myCRAN" = "file:///.../myCRAN/", "CRAN" = "https://cran.rstudio.com/")
options(repos = repos)
```
When I try to install the package from the repo on local R it works.
```R
install.packages("...something...", type = "source")
```
When I try to deploy on Rsconnect it fails:
```
2018/01/24 13:24:37.867498014 Installing ... 
2018/01/24 13:24:37.950072827 curl: (37) Couldn't open file /....tar.gz
2018/01/24 13:24:39.003153892 curl: (37) Couldn't open file /....tar.gz
2018/01/24 13:24:40.023737512 curl: (37) Couldn't open file /....tar.gz
2018/01/24 13:24:41.043526005 curl: (37) Couldn't open file /....tar.gz
2018/01/24 13:24:42.064211526 curl: (37) Couldn't open file /....tar.gz
2018/01/24 13:24:43.641243092 curl: (22) The requested URL returned error: 404 Not Found
2018/01/24 13:24:45.037615828 curl: (22) The requested URL returned error: 404 Not Found
2018/01/24 13:24:46.361135668 curl: (22) The requested URL returned error: 404 Not Found
2018/01/24 13:24:47.623601503 curl: (22) The requested URL returned error: 404 Not Found
2018/01/24 13:24:48.798805636 curl: (22) The requested URL returned error: 404 Not Found
2018/01/24 13:24:49.802007705 FAILED
2018/01/24 13:24:49.809510696 Error in getSourceForPkgRecord(pkgRecord, srcDir(project), availablePackagesSource(repos = repos), : Failed to retrieve package sources for ...
```
It does that on the very first package it tries to install (which is available on Cran.
This workaround works:
I declare in Rstudio connect that my private package is external and is installed on the server with Rstudio connect.
I remove any declaration to a local host in my options() file.
Then I can deploy.

I think there is a bug with rsconnect client and local Cran when the address is provided as local path.

**Comment by trestletech** [comment]
Hi.  As an RStudio Connect customer I'd encourage you to send in requests to support@rstudio.com where we have folks devoted specifically to answering these questions. GitHub responses may be slower at times.

Do you mind sending this there so we can make sure that we properly log the bug report on Connect if need be?

Thanks!

**Comment by pommedeterresautee** [comment]
Already did it (request 25193) and already linked it to that issue.
However, it seems to be an issue with the client `rsconnect`, that s why I opened also here.
Other may be interested if the solution is a workaround for instance.

**Comment by kevinushey** [comment]
Any chance we can get access to the bundle generated by `rsconnect` during deployment?

**Comment by cderv** [comment]
For inspecting behavior before deployement in RStudio Connect, I found `rsconnect:::performPackratSnapshot(".")` very useful. This is a non exported function used internally in a temp directory but if you run it in the folder to deploy the app, it will generate a packrat folder with the lock file that would be use in RStudio Connect during deployement. 
This `packrat.lock` file helped me in the past to understand what is going on with repository and dependencies. It may help understand.

The `manifest.json` could be useful to but it is created during the bundle so last request above applies.

**Comment by pommedeterresautee** [comment]
@cderv the lock file generated seems Ok to me, I see in particular my private package marked as "myCRAN" in `source` key.

```
...other packages...

Package: RcppParallel
Source: CRAN
Version: 4.3.20
Hash: af8415c361576a8c4c5ce7ad344da3fc
Requires: BH

Package: SOMETHING
Source: myCRAN
Version: 0.0.1.9000
Hash: 3fcca9f27a5ff2cf6a2b31d53d1c8216
Requires: SOMETHING, SOMETHING, SOMETHING
...
```


@kevinushey how can I retrieve the bundle?

My deploy script look like that:
```R
cran_path <- file.path(getwd(), "myCRAN")
repos <- c("myCRAN" = paste0("file://", cran_path), "CRAN" = "https://cran.rstudio.com/")
options(repos = repos)
getOption("repos")

# it works and the package is only available on my local Cran
install.packages("SOMETHING", type = "source")

rsconnect::deployApp(
  appDir = "./webapps/SOMETHING/",
  appFiles = c('app.R'),
  appName = "CI-Deployed"
)
```

The results from the first line:
```
> rsconnect::deployApp(
+   appDir = "./webapps/SOMETHING/",
+   appFiles = c('app.R'),
+   appName = "CI-Deployed"
+ )
Preparing to deploy application...
Update application currently deployed at
http://127.0.0.1:3939/content/1/? [Y/n] Y
DONE
Uploading bundle for application: 1...DONE
Deploying bundle: 3 for application: 1 ...
Building Shiny application...
Bundle requested R version 3.4.3; using /usr/lib/R/bin/R which has version 3.4.3
2018/01/25 07:21:22.845999224 # Validating R library read / write permissions --------------------------------
2018/01/25 07:21:22.884596637 # Validating packrat installation ----------------------------------------------
2018/01/25 07:21:22.884650081 Installed packrat SHA is: 4b3e4025404f98d25fb73799e86c1762effe5d7e
2018/01/25 07:21:22.884686097 Packaged packrat SHA is:  4b3e4025404f98d25fb73799e86c1762effe5d7e
2018/01/25 07:21:22.885516968 Packrat is up-to-date.
2018/01/25 07:21:22.885599874 R version: 3.4.3
2018/01/25 07:21:22.885608552 packrat version: 0.4.8.52
2018/01/25 07:21:22.886028379 # Validating packrat cache read / write permissions ----------------------------
2018/01/25 07:21:23.007922353 Audited package hashes with local packrat installation.
2018/01/25 07:21:23.010527006 # Installing required R packages with `packrat::restore()` ---------------------
2018/01/25 07:21:23.216917092 Installing BH (1.65.0-1) ... 
2018/01/25 07:21:23.289386158 curl: (37) Couldn't open file /mnt/workspace/elia/myCRAN/src/contrib/Archive/BH/BH_1.65.0-1.tar.gz
2018/01/25 07:21:24.332382019 curl: (37) Couldn't open file /mnt/workspace/elia/myCRAN/src/contrib/Archive/BH/BH_1.65.0-1.tar.gz
2018/01/25 07:21:25.352240854 curl: (37) Couldn't open file /mnt/workspace/elia/myCRAN/src/contrib/Archive/BH/BH_1.65.0-1.tar.gz
2018/01/25 07:21:26.372758314 curl: (37) Couldn't open file /mnt/workspace/elia/myCRAN/src/contrib/Archive/BH/BH_1.65.0-1.tar.gz
2018/01/25 07:21:27.390574917 curl: (37) Couldn't open file /mnt/workspace/elia/myCRAN/src/contrib/Archive/BH/BH_1.65.0-1.tar.gz
2018/01/25 07:21:28.611497032 curl: (22) The requested URL returned error: 404 Not Found
2018/01/25 07:21:29.746030789 curl: (22) The requested URL returned error: 404 Not Found
2018/01/25 07:21:30.850586757 curl: (22) The requested URL returned error: 404 Not Found
2018/01/25 07:21:31.984393954 curl: (22) The requested URL returned error: 404 Not Found
2018/01/25 07:21:33.086493774 curl: (22) The requested URL returned error: 404 Not Found
2018/01/25 07:21:34.089982281 FAILED
2018/01/25 07:21:34.095118127 Error in getSourceForPkgRecord(pkgRecord, srcDir(project), availablePackagesSource(repos = repos), : Failed to retrieve package sources for BH 1.65.0-1 from CRAN (internet connectivity issue?)
2018/01/25 07:21:34.095144173 
2018/01/25 07:21:34.095185896 Unable to fully restore the R packages associated with this deployment.
2018/01/25 07:21:34.095193609 Please review the preceding messages to determine which package
2018/01/25 07:21:34.095211805 encountered installation difficulty and the cause of the failure.
2018/01/25 07:21:34.095213515 
2018/01/25 07:21:34.095229465 Some typical reasons for package installation failures:
2018/01/25 07:21:34.095231135   * A system library needed by the R package is not installed.
2018/01/25 07:21:34.095246951     Some of the most common package dependencies are cataloged at:
2018/01/25 07:21:34.095248518     https://github.com/rstudio/shinyapps-package-dependencies
2018/01/25 07:21:34.095263981 
2018/01/25 07:21:34.095265584   * The R package requires a newer version of R.
2018/01/25 07:21:34.095288138 
2018/01/25 07:21:34.095290077   * The C/C++ compiler is outdated. This is often true for packages
2018/01/25 07:21:34.095306030     needing C++11 features.
2018/01/25 07:21:34.095307777 
2018/01/25 07:21:34.095323227   * The R package is Windows-only or otherwise unavailable for this
2018/01/25 07:21:34.095324810     operating system.
2018/01/25 07:21:34.095361936 
2018/01/25 07:21:34.095363846   * The package is housed in a private repository that requires
2018/01/25 07:21:34.095379629     authentication to access. For more details on this, see:
2018/01/25 07:21:34.095381306     http://docs.rstudio.com/connect/admin/process-management.html#private-packages
2018/01/25 07:21:34.095403356 
2018/01/25 07:21:34.095405216 The package description and documentation will list system requirements
2018/01/25 07:21:34.095421135 and restrictions.
2018/01/25 07:21:34.095422802 
2018/01/25 07:21:34.095438028 Please contact your RStudio Connect administrator for further help
2018/01/25 07:21:34.095439645 resolving this issue.
2018/01/25 07:21:34.101009741 Warning message:
2018/01/25 07:21:34.101025788 In read.dcf(file = tmpf) :
2018/01/25 07:21:34.101062560   cannot open compressed file '/mnt/workspace/elia/myCRAN/src/contrib/PACKAGES', probable reason 'No such file or directory'
Application deployment failed with error: exit status 1
```

For what it worths:
```
> getOption("repos")
                             myCRAN                                CRAN 
"file:///mnt/.../myCRAN"         "https://cran.rstudio.com/" 
```

**Comment by cderv** [comment]
Thanks for the input. The repository declarations are at the beginning of the lockfile. You may have checked it already. 

What seems odd is that ˋBH` is trying to get installed from your local myCRAN not from the other. Is this normal? Can you confirm that ˋ/mnt/workspace/elia/myCRAN/src/contrib/Archive/BH/BH_1.65.0-1.tar.gz` does not exist locally on myCRAN. I understand that your drat repo contains only you github private package. 
You may check in the lockfile if the CRAN dependencies for your private package are indicated to download from CRAN or myCRAN. 

For getting the bundle app, you can use ˋrsconnect:::bundleApp` with the right arguments. 
Or deploy in debug mode and look into your ˋtempdir()` folder. 

I encountered some issue in the past with some edge cases between rsconnect and packrat. It is some work to isolate the real issue.

**Comment by pommedeterresautee** [comment]
Regarding the order of installation, it is because I have declared my repo in this order : 
```R
repos <- c("myCRAN" = paste0("file://", cran_path), "CRAN" = "https://cran.rstudio.com/")
```
It was one of my experience (I am desperate enough to try this kind of ideas).
Of course when I reverse the order I get first the 404 and then the couldn't open file.

The Drat repo contains only the private package and nothing else. I have rebuilt it many times, and checked the `PACKAGES` index which seems right.

An extract of the lockfile is provided in the message above, it s the SOMETHING line (I have just changed the name because the conversion is public).
It is marked as myCRAN, otherwise, I have tried to also no include the Repo marker in `DESCRIPTION` file, it refuses to start the upload of the bundle.

Thank you for the retrieve bundle methodology, I will try that.

**Comment by kevinushey** [comment]
For context, this is the codepath Packrat is executing when attempting to restore the package:

https://github.com/rstudio/packrat/blob/3eb9c1250ae4384012164a75282f6992c4aa972e/R/restore.R#L183-L198

And that `repos` is typically inherited from the lockfile, as per:

https://github.com/rstudio/packrat/blob/3eb9c1250ae4384012164a75282f6992c4aa972e/R/packrat.R#L396-L403

Based on the output, it _seems_ like, for some reason, Packrat isn't ever trying to restore BH from the global CRAN repository -- which would imply that the repository is somehow getting lost somewhere.

**Comment by kevinushey** [comment]
It might be helpful to know the output of

```
packrat::snapshot(snapshot.sources = FALSE)
packrat:::lockInfo('.', 'repos')
```

**Comment by pommedeterresautee** [comment]
Ok I understand what is happening.
I point to a local repo that is not reachable by Rstudio connect.
Rsconnect client doesn't check that, it just say that my custom package is associated with the custom repo.
When the first package is going to be installed, rstudio connect, for some reasons probably related to the fact that one repo is not reachable, becomes crazy, and fail installation of the first package. I know that it is at that moment it becomes crazy because if I declare a fake online repo with the name myCran it fails at the last package, the private one, but BH and the others are installed (and if I configure Rstudio connect to use system packages, and my private package is already installed, it just works as said before).

My current workaround: I serve the local repo with a server and I declare only the IP and the port, but not a local repo not reachable by Rstudio connect. So Rstudio connect has no more issue.
I do that in r directly:

-------
cran_path <- file.path(getwd(), "myCran")
ip <- "172.17.0.1"
port <- 4040
clone_path <- "./test_clone"
backup_repo <- getOption("repos")

unlink(clone_path, recursive = TRUE)
git2r::clone("git@github.com:xyz.git", clone_path, progress = FALSE)
# Build local Cran
unlink(cran_path, recursive = TRUE)
package_source_path <- devtools::build(clone_path, vignettes = FALSE)
unlink(clone_path, recursive = TRUE)
dir.create(cran_path)
drat::insertPackage(package_source_path, cran_path)
drat::addRepo(account = "ELSCran", alturl = paste0("http://", ip, ":", port))
unlink(package_source_path)

# provide local Cran as a webservice
servr::httd("./ELSCran/", host = ip, port = port, daemon = TRUE, browser = FALSE)

# test Cran repo and update the app in case it is not already the case
install.packages("...", type = "source")

# Deploy application on Rstudio connect
unlink("./webapps/.../rsconnect/", recursive = TRUE)
rsconnect::deployApp(
appDir = "./webapps/.../",
appFiles = c('app.R')
)

# stop daemon
servr::daemon_stop(which = servr::daemon_list())
# restore repo
options(repos = backup_repo)
-------

May be some of your users would be interested in such workaround for local Cran repo.
Of course, the workaround is not compatible with a CI deployment :-(

I have understood that a feature to support private git was coming to Rstudio-connect.
Do you have an idea of when it will happen? (I am not asking for a strong dead line but just to have a basic idea so I can plan a switch to Travis deployment).

**Comment by cderv** [comment]
Nice workaround! As it is a prerequisite that your rstudio connect server has direct access to cran repo, I assume this was not the issue. 

Your solution corresponds to setting a _temporary_ private repository. Don’t you have the ability to have a _persistent_ cran repos (web server for example) where you can publish your private packages (using CI) and that will be accessible from rstudio connect server? 
It is what we are using for this use case.

**Comment by pommedeterresautee** [comment]
I can of course create a Cran server, but it would mean add a new server to our infra, plus add a new task to the CI (push new package to Cran server) which is not that easy to build (how do you update private Cran repo with a simple server, you need something active.
And more important it breaks the logic/USP of Rstudio connect, the Shiny app or whatever is supposed to be pushed from one point, all the package are supposed to be the one the dev has used during his dev, and the process not being too complex. Distant Cran means potentially another version of the package, a complex process relying on different sources.

**Comment by jmcphers** [comment]
After the investigation above, I don't think there are changes we can or should make in the package at this point. (Let me know if you disagree!)
